project_euler/problem_203/sol1.py

```diff
@@ -0,0 +1,181 @@
+in the first eight rows is 105.
+
+Find the sum of the distinct squarefree numbers in the first 51 rows of
+Pascal's triangle.
+
+References:
+- https://en.wikipedia.org/wiki/Pascal%27s_triangle
+"""
+
+import math
+from typing import List, Set
+
+
+def get_pascal_triangle_unique_coefficients(depth: int) -> Set[int]:
+    """
+    Returns the unique coefficients of a Pascal's triangle of depth "depth".
+
+    The coefficients of this triangle are symmetric. A further improvement to this
+    method could be to calculate the coefficients once per level. Nonetheless,
+    the current implementation is fast enough for the original problem.
+
+    >>> get_pascal_triangle_unique_coefficients(1)
+    {1}
+    >>> get_pascal_triangle_unique_coefficients(2)
+    {1}
+    >>> get_pascal_triangle_unique_coefficients(3)
+    {1, 2}
+    >>> get_pascal_triangle_unique_coefficients(8)
+    {1, 2, 3, 4, 5, 6, 7, 35, 10, 15, 20, 21}
+    """
+    coefficients = {1}
+    previous_coefficients = [1]
+    for step in range(2, depth + 1):
+        coefficients_begins_one = previous_coefficients + [0]
+        coefficients_ends_one = [0] + previous_coefficients
+        previous_coefficients = []
+        for x, y in zip(coefficients_begins_one, coefficients_ends_one):
+            coefficients.add(x + y)
+            previous_coefficients.append(x + y)
+    return coefficients
+
+
+def get_primes_squared(max_number: int) -> List[int]:
+    """
+    Calculates all primes between 2 and round(sqrt(max_number)) and returns
+    them squared up.
+
+    >>> get_primes_squared(2)
+    []
+    >>> get_primes_squared(4)
+    [4]
+    >>> get_primes_squared(10)
+    [4, 9]
+    >>> get_primes_squared(100)
+    [4, 9, 25, 49]
+    """
+    max_prime = round(math.sqrt(max_number))
+    non_primes = set()
+    primes = []
+    for num in range(2, max_prime + 1):
+        if num in non_primes:
+            continue
+
+        counter = 2
+        while num * counter <= max_prime:
+            non_primes.add(num * counter)
+            counter += 1
+
+        primes.append(num ** 2)
+    return primes
+
+
+def get_squarefree(
+    unique_coefficients: Set[int], squared_primes: List[int]
+) -> Set[int]:
+    """
+    Calculates the squarefree numbers inside unique_coefficients given a
+    list of square of primes.
+
+    Based on the definition of a non-squarefree number, then any non-squarefree
+    n can be decomposed as n = p*p*r, where p is positive prime number and r
+    is a positive integer.
+
+    Under the previous formula, any coefficient that is lower than p*p is
+    squarefree as r cannot be negative. On the contrary, if any r exists such
+    that n = p*p*r, then the number is non-squarefree.
+
+    >>> get_squarefree({1}, [])
+    set()
+    >>> get_squarefree({1, 2}, [])
+    set()
+    >>> get_squarefree({1, 2, 3, 4, 5, 6, 7, 35, 10, 15, 20, 21}, [4, 9, 25])
+    {1, 2, 3, 5, 6, 7, 35, 10, 15, 21}
+    """
+
+    def get_squared_primes_to_use(
```

Comment:
A function within a function is mostly used as a wrapper. This doesn't seem like that so please separate them out.

Reply:
Done, thanks for the advice!

Taking out the function also revealed that one doctest of `get_squared_primes_to_use` was failing. I fixed it and revisited the function to gracefully handle that failing case.